From fe3d9e5abd4538186ec489c6e16d910cb6d657c0 Mon Sep 17 00:00:00 2001
From: pedro <pedro311@gmail.com>
Date: Sun, 19 Jul 2026 19:07:59 +0200
Subject: [PATCH 1/3] ntpd: avoid undefined shifts for peer precision

m_precision_exp is an int8_t received from the network.  LOG2D()
currently shifts an unsigned long by the absolute value of that field.
Exponents outside the native word width therefore trigger undefined
behavior, e.g. 1UL << 64 on a 32-bit target.

Scale in chunks of 30 bits so every shift is valid while retaining the
full range of the protocol field.

Signed-off-by: pedro311 <pedro311@gmail.com>
---
 busybox/networking/ntpd.c | 18 ++++++++++++++++--
 1 file changed, 16 insertions(+), 2 deletions(-)

diff --git a/networking/ntpd.c b/networking/ntpd.c
index 4d48f3ddc..49bf0d863 100644
--- busybox/networking/ntpd.c
+++ busybox/networking/ntpd.c
@@ -480,9 +480,23 @@ struct globals {
 
 static double LOG2D(int a)
 {
+	double d = 1.0;
+
+	/*
+	 * The precision exponent comes from the network and is an int8_t.
+	 * Keep shifts below the width of unsigned long on 32-bit targets.
+	 */
+	while (a > 30) {
+		d *= (1UL << 30);
+		a -= 30;
+	}
+	while (a < -30) {
+		d /= (1UL << 30);
+		a += 30;
+	}
 	if (a < 0)
-		return 1.0 / (1UL << -a);
-	return 1UL << a;
+		return d / (1UL << -a);
+	return d * (1UL << a);
 }
 static ALWAYS_INLINE double SQUARE(double x)
 {
-- 
2.39.5

